### DISCUSSION TITLE
GoogleMapsOverlay, MVTLayer, and memory

### DISCUSSION THREAD

**Discussion by dCogs** [visgl/deck.gl#5524] [Q&A] [answered]
Thanks to all of the deck.gl contributors out there. I'm finding it to be the best option for working with large datasets on Google Maps.

I use @deck.gl/google-maps to load MVTLayers onto Google Maps. Usually it's a single layer on a GoogleMapsOverlay instance. This creates several (usually 5) instances of **mapbox vector tile-worker@2.3.11**, each with several MBs of memory (one exceeded 10MB). 

When I am done with a layer, I `finalize()` the GoogleMapsOverlay instance after performing `finalizeState()` on each layer. After letting javascript do its garbage collection, I can see that the main heap size has gone back to its normal size, however the 5 instances of **mapbox vector tile-worker@2.3.11** remain in memory regardless of how long I wait. When looking at memory in web developer and I mouse over the instance, it says "_Heap size in use by live JS objects_"

I'm curious if this is expected behavior.

Many thanks,
Dan

**Comment by ibgreen** [comment, marked as answer]
loaders.gl (which deck.gl uses under the hood) offers options to control the number of worker threads and also there is a `reuseWorkers: false` option.

I believe that these are only read the first time a worker farm is created, so you need to provide them in the first `load` or `parse` call, or set them with deck's load options or `setLoaderOptions`

I don't have any examples handy but maybe these links will give you enough of a clue to get it working:

https://github.com/visgl/loaders.gl/blob/1c12b3e743d05eecd9737387bdf15c0095786de8/modules/worker-utils/src/lib/worker-farm/worker-farm.js#L27

https://github.com/visgl/loaders.gl/blob/5253b009b919277b1f548d4e7467546ce0469600/modules/core/src/lib/loader-utils/parse-with-worker.js#L36

**Reply by dCogs** [reply]
That did it, thanks. A slightly noticeable decline in responsiveness when zooming in/out, but the this removes the extra threads when the MVT layer is no longer used.

```
loadOptions: {
    reuseWorkers: false
}
```

**Reply by ibgreen** [reply]
Great. I feel the default of creating 5 workers per loader is excessive. I see diminishing returns beyond 2 workers even on systems with more cores. Considering reducing the number of default workers.

On mobile it may be desirable to reduce it further to a single worker to minimize memory pressure.

Maybe loaders.gl could also provide an explicit API to terminate workers once loading is done?
